### ISSUE TITLE
Integrate dictionaries

### ISSUE THREAD

**Issue by RustanHakansson** [lexica/lexica#187]
Currently it is possible to look up a word with AARD2, QuickDic or DuckDuckGo.

AARD2 is quite hard to get the definitions for. Installing the app and trying to download the dictionary for Swedish leads me to a webpage, a github page where the first is a magnet link that has no seeders, the second to a 44mb file that works. 

QuickDict is easier to download dictionary for, but is really annoying as it requires an additional click to open the word from a list of closest matches, then after reading I need to go back twice. This makes me refrain from looking up words, as it is too cumbersome. 

Actually getting a definition from the missed words list is quite tricky, I often need several tries to hit the icon just the way to app wants it. I have not found a reliable way to get it to open, and have no other issues like this elsewhere in the app. 

I tested the other dictionary apps on F-Droid, and really like what "DictionaryForMIDs" does. http://dictionarymid.sourceforge.net/what.html 
You download the dictionary from within the app itself very easily, no other website to go to. The definitions list when using data from OmegaWiki http://omegawiki.org/  is short and easy to read, and would fit well for use in the missed words list in the app. 

For example when searching for "tree" it has definitions in these languages:

```
Breton | Plant bras koadek dezho ur c'hef dalc'hus a sav skourroù pe delioù dioutañ.
-- | --
Bulgarian | Всяко голямо многогодишно дървесно растение със ствол, от който тръгват разклонения или листа на известно разтояние от земята.
Castilian | Planta perenne arbolada grande con un tronco distinto que da lugar a ramas o a hojas a una cierta distancia de la tierra.
Dutch | Een houtachtige, meerjarige plant met een duidelijke stam die zich op enige hoogte boven de grond vertakt.
English | Any large woody perennial plant with a distinct trunk giving rise to branches or leaves at some distance from the ground.
French | Grande plante ligneuse avec un tronc duquel partent des branches ou des feuilles.
German | Große, mehrjährige Pflanze mit einem ausgeprägten Stamm aus dem ab einer gewissen Höhe Äste und Blätter wachsen.
Italian | Grande  pianta perenne legnosa con un tronco ben visibile a cui sono collegati i  rami e le foglie che si ergono ad una certa distanza dal suolo.
Mandarin (simplified) | 任一拥有明显的树干在离地有一定距离处带有树枝或叶子的大型多年生植物。
Polish | Każda  roślina wieloletnia dużych rozmiarów o wyraźnie wykształconym pniu,  który w pewnej wysokości nad ziemią rozgałęzia się w koronę
Portuguese | Vegetal grande, de tronco robusto, que se reveste de folhas
Russian | Большое деревянное растение с чётким стволом которым имеет ветви или листья в какое-то положение над землёй.
Slovenian | Vse velike lesene trajnice z izrazitim deblom iz katerega na neki oddaljenosti od tal rastejo veje ali listi.
Vietnamese | Cá thể thực vật, thường có thân, lá, cành.
```

If such a short explanation could be shown in the app itself, with backup to go to another provider when nothing is found, it would make word learning much faster and more fun. 

OmegaWiki is an offshoot from Wiktionary, using the same tools and same policies to a large extent, but stores single concepts in a single page instead of in one per language: http://www.omegawiki.org/Help:Advantages_of_OmegaWiki_over_Wiktionary

Definition files are really clean and simple, just a csv with one column for word and one for definition: http://omegawiki.org/Special:Ow_downloads
If there are multiple uses of a word, there are multiple lines. 

There are only stems, so "trees" would need to look up "tree" to get a definition. 

The English definition file has over 5M entries, but still only 5 mb. Here are the languages with the most definitions:

```
Breton | 216898
-- | --
Bulgarian | 938634
Castilian | 4235927
Dutch | 1707900
English | 5279644
French | 2600484
German | 1778355
Greek | 50208
Interlingua | 103329
Italian | 2700061
Japanese | 36756
Kölsch | 199058
Piedmontese | 58032
Polish | 447894
Russian | 53840
Slovenian | 544544
Swedish | 188880
Turkish | 220691
```

Integrating the definitions would make it much easier for users to use them, instead of relying on an external app. As the files are so well-defined, it might be possible to download the definitions as needed when the user changes to a new language. This would make it easier to maintain, as the definitions do not need to be part of the repo. 

Wiktionary has a lot more definitions, but they are a lot longer and more complex. I cannot find any files comparable to the above ones on omegawiki.

**Comment by pserwylo** [comment]
As evidenced by #186, having the ability to show definitions within the app would be wonderful. There are two things that make me take a step back from this though:

1. Maintaining different wordlists is a top priority (required to play the game in various languages). However as seen, [there are many issues that arise when trying to maintain multiple wordlists](https://github.com/lexica/lexica/labels/dictionaries), many of which have not been resolved for some time due to resource constraints. This would double (or perhaps more) the backlog / work involved, which takes effort away from other improvements (e.g. our desire for multiple game modes).

2. As you have mentioned, there is a non-zero size overhead for this. 5MiB per language would be quite large, on top of the dictionary sizes. *However*, I'm less worried about this. For example, assets could be hosted online and downloaded when required (e.g. "Download Swedish Dictionary" would then fetch the assets the first time a user wanted them).

**Comment by RustanHakansson** [comment]
I fully understand. My main goal in looking for word lists, letter distributions and dictionaries have been to try to find such in common formats that can be set up in bulk, as well as updated in bulk, and not need them to be something that time is spent on. 

Letter distribution #184 might be a one-off, so not really problematic to store here if done in bulk for 50 languages. 

OmegaWiki might be a useable solution where the files can be downloaded easily directly from that site, for a large group of languages. 

Word lists might be trickier, and the thing that needs to be kept in the repo, like now, unless a common external repo can be used (like tube42 word list in #179 ). Using Wiktionary or other similar systems is a lot trickier for this than for dictionaries. I do not think dictionaries need to be treated the same as word lists.

**Comment by pserwylo** [comment]
Hmm, I wonder if a temporary solution to allow in-game definitions would be to use a `WebView` in a popup dialog? This could be done right now with very little effort.

**Comment by RustanHakansson** [comment]
Yes, that would be great! Could be used for the long list as well. It is not a core functionality, the app is still useable offline and if you have an offline dictionary you can look up words yourself. Having the definitions stored locally is not critical, it is much better to have a really useful default setup without having to install and configure an additional app. 

It would be a bit of work to strip the webview down to only show the relevant data, but that is hopefully work that only needs to be done once per site used, not once per language. 

Perhaps make it possible to choose between omegawiki and wiktionary, as some would prefer a shorter and some a longer definition. Maybe even more if there are languages better supported by other open sites. 

Hmm, thinking a bit more about this, I suggest checking out the apis they provide instead:

http://www.omegawiki.org/Help:OmegaWiki_API
https://en.wiktionary.org/w/api.php

My guess is that this is easier and faster than messing with a webview, and uses a lot less resources both on device and on sites.

**Comment by GrimPixel** [comment]
Good to see this. It would be much more helpful if the meaning appears when the word is revealed.

Dictionaries can be easily found here:
https://polyglotclub.com/wiki/Language/Multiple-languages/Culture/Internet-Dictionaries
